File: lib/src/main/java/io/github/kory33/guardedqueries/core/subsumption/formula/IndexlessMaximallySubsumingTGDSet.java
package io.github.kory33.guardedqueries.core.subsumption.formula;

import com.google.common.collect.ImmutableSet;
import uk.ac.ox.cs.pdq.fol.TGD;

import java.util.HashSet;

/**
 * A partial implementation of {@link MaximallySubsumingTGDSet} that does not use any index
 * to filter out subsumption candidates.
 */
public abstract class IndexlessMaximallySubsumingTGDSet<F extends TGD> implements MaximallySubsumingTGDSet<F> {
    /**
     * Judge if the first rule subsumes the second rule,
     * according to an implementation-specific subsumption relation.
     */
    protected abstract boolean firstRuleSubsumesSecond(F first, F second);

    private final HashSet<F> rulesKnownToBeMaximalSoFar = new HashSet<>();

    private boolean isSubsumedByExistingRule(F rule) {
        return rulesKnownToBeMaximalSoFar.stream()
                .anyMatch(existingRule -> firstRuleSubsumesSecond(existingRule, rule));
    }

    @Override
    public final void addRule(F rule) {
        if (!isSubsumedByExistingRule(rule)) {
            rulesKnownToBeMaximalSoFar.removeIf(existingRule -> firstRuleSubsumesSecond(rule, existingRule));
            rulesKnownToBeMaximalSoFar.add(rule);
        }
    }

    @Override
    public final ImmutableSet<F> getRules() {
        return ImmutableSet.copyOf(rulesKnownToBeMaximalSoFar);
    }
}

File: lib/src/main/scala/io/github/kory33/guardedqueries/core/subsumption/formula/IndexlessMaximallySubsumingTGDSet.scala
package io.github.kory33.guardedqueries.core.subsumption.formula

import uk.ac.ox.cs.pdq.fol.TGD
import java.util

/**
 * A partial implementation of [[MaximallySubsumingTGDSet]] that does not use any index to
 * filter out subsumption candidates.
 */
abstract class IndexlessMaximallySubsumingTGDSet[F <: TGD] extends MaximallySubsumingTGDSet[F] {

  /**
   * Judge if the first rule subsumes the second rule, according to an implementation-specific
   * subsumption relation.
   */
  protected def firstRuleSubsumesSecond(first: F, second: F): Boolean
  final private val rulesKnownToBeMaximalSoFar = new util.HashSet[F]
  private def isSubsumedByExistingRule(rule: F) = rulesKnownToBeMaximalSoFar.stream.anyMatch(
    (existingRule: F) => firstRuleSubsumesSecond(existingRule, rule)
  )
  override final def addRule(rule: F): Unit = {
    if (!isSubsumedByExistingRule(rule)) {
      rulesKnownToBeMaximalSoFar.removeIf((existingRule: F) =>
        firstRuleSubsumesSecond(rule, existingRule)
      )
      rulesKnownToBeMaximalSoFar.add(rule)
    }
  }
  override final def getRules: Set[F] = Set.copyOf(rulesKnownToBeMaximalSoFar)
}
